Fixes for x86 LIR

BruceForstall opened the pull request:
With this, crossgen of x86 S.P.C.dll is as clean as the non-LIR version. There is a known
assert with checked. With debug, there are a number of crashes with JMP calls that also
occur with the legacy JIT.

There is one big caveat: I added an NYI in LowerArg for long stack arguments. This hits 450 times.
The code is:

```
// TODO(pdg): the code below uses tree order to freely reorder the lo and hi pieces of
// a long-type argument. This needs to be fixed.
NYI("LIR: long arg lowering");
```

This will be fixed next.

The changes:
1. During decomposition, make sure tree costs are set.
2. Fix bug in LIR::Range::rend() if the range is empty.
3. In CheckLIR(), make sure the forward and reverse threadings have no circularities.

Also:
a. Dump MethodHash with COMPLUS_JitFunctionTrace

@pgavlin PTAL

BruceForstall committed:
Fixes for x86 LIR

With this, crossgen of x86 S.P.C.dll is as clean as the non-LIR version. There is a known
assert with checked. With debug, there are a number of crashes with JMP calls that also
occur with the legacy JIT.

There is one big caveat: I added an NYI in LowerArg for long stack arguments. This hits 450 times.
The code is:
```
// TODO(pdg): the code below uses tree order to freely reorder the lo and hi pieces of
// a long-type argument. This needs to be fixed.
NYI("LIR: long arg lowering");
```
This will be fixed next.

The changes:
1. During decomposition, make sure tree costs are set.
2. Fix bug in LIR::Range::rend() if the range is empty.
3. In CheckLIR(), make sure the forward and reverse threadings have no circularities.

Also:
a. Dump MethodHash with COMPLUS_JitFunctionTrace

pgavlin commented on src/jit/lir.cpp:
```diff
@@ -1110,6 +1113,28 @@ bool LIR::Range::CheckLIR(Compiler* compiler, bool checkUnusedValues) const
     // Make sure the range itself is valid.
     assert(IsValid());
 
+    // Make sure the list is well-formed: no node appears more than once (no circularities). Insert each node in a set
```
Did you actually run in to circularities? It shouldn't be possible to construct such lists using the `Insert*`/`Remove` methods (though it is possible via manual mutations of `gtNext`/`gtPrev`)

BruceForstall replied:
Yes. I think if you call Insert*(SeqTree()) on a tree that points to existing LIR nodes this can happen.

pgavlin replied:
As mentioned below, SeqTree should never be called on a tree that points to existing LIR. Unfortunately I'm not quite sure how to write an assert that catches that :(

pgavlin commented on src/jit/lir.cpp:
```diff
@@ -1110,6 +1113,28 @@ bool LIR::Range::CheckLIR(Compiler* compiler, bool checkUnusedValues) const
+    // Make sure the list is well-formed: no node appears more than once (no circularities). Insert each node in a set
+    // and assert we don't find the node again.
+    typedef SimplerHashTable<GenTreePtr, PtrKeyFuncs<GenTree>, bool, JitSimplerHashBehavior> NodeToBoolMap;
+    NodeToBoolMap* nodes = new (compiler->getAllocatorDebugOnly()) NodeToBoolMap(compiler->getAllocatorDebugOnly());
```
Instead of using a map, this could use the LIR mark bit to track visited nodes. It's not critical, but it will save some  space (and potentially time, as the forward scan could be incorporated into the later checks). You could check that the bit is not set in the forward scan and that the bit is set in the reverse scan.

BruceForstall replied:
The problem is, you need to first clear all the mark bits, which requires iterating over the links you're checking for legality. Also, I generally prefer that debug functions don't mutate the things they're checking.

I'll change this to the two pointer, fast/slow algorithm.

Another property I want to check is that the gtPrev list is precisely the opposite of the gtNext list.

pgavlin commented on src/jit/decomposelongs.cpp:
```diff
@@ -796,11 +799,9 @@ GenTree* DecomposeLongs::DecomposeInd(LIR::Use& use)
-    m_blockRange.InsertAfter(addrBaseHigh, indLow);
-    m_blockRange.InsertAfter(addrHigh, addrBaseHigh);
-    m_blockRange.InsertAfter(indHigh, addrHigh);
+    m_blockRange.InsertAfter(LIR::SeqTree(m_compiler, indHigh), indLow);
```
Calling `SeqTree` here is probably a bit less efficient than just manually inserting the nodes and adds some debt for later changes that I'd like to make s.t. we are able to do shallow cost calculation on LIR nodes. Did you switch to using this because it initializes cost information or for the sake of brevity? If it is the former, I'd prefer that we just call `gtPrepareCost` and leave the manual insertion as-is. Same goes for the similar changes in `DecomposeStoreInd`.

pgavlin replied:
And just to be clear, what I've suggested w.r.t. calling `gtPrepareCost` before manually inserting the nodes will be a bit less efficient than the change you've made, but I think it gives us a clear marker for where we'll need to update the code to do shallow cost calculation.

pgavlin replied:
After some more thought, I've remembered that there's a deeper problem with calling `SetTreeSeq` on a tree that contains nodes tht are already in an LIR block: the implementation calls `fgSetTreeSeq`, which mutates `gtNext`/`gtPrev` for all nodes in the tree. This is very much not safe for nodes that are already in a block, as blindly resequencing such nodes may perform incorrect code motion. IIRC I even added a comment to this effect in `SetTreeSeq` as part of #24.

BruceForstall replied:
Right, but for the cases I converted to SeqTree(), they are not already in the block. The cases I didn't convert to SeqTree do refer to existing nodes.

pgavlin replied:
I don't think that's correct. There are two cases:
1. All operands have been replaced with local variables using `ReplaceWithLclVar`
2. Some/all operands have not been replaced with local variables

In the first case, the local var use that is created to replace the use of the spilled tree is threaded into the LIR block immediately after the store of the spilled tree. See the comment above `ReplaceWithLclVar` for example IR. In this case it should be sufficient to remove the local var uses from the block before calling `SeqTree` as these variables are guaranteed to be unaliased.

In the second case, the operands must be present in the LIR block by construction.  In this case, it is neither sufficient nor necessarily safe to remove the operands from the LIR block before calling `SeqTree`, as sequencing the tree and then inserting the new range may have moved the operands across nodes with conflicting side effects.

In both cases, I think it is simpler and safer to just insert the new IR manually.

BruceForstall replied:
There's a third case: 3. a new local var ref is created to an operand.

In the StoreInd case, the operand is removed from the range before the new tree is sequenced, then re-inserted by the InsertAfter code ("code motion" of the local/leaf), so SeqTree would be safe. In the DecomposeInd case, there is also no sharing, as the "addrBaseHigh" creates a new GT_LCL_VAR reference.

I'm ok using gtPrepareCosts + a sequence of node InsertAfter, but it's more lines of code and less easy to read/reason about, IMO.

pgavlin replied:
> In the StoreInd case, the operand is removed from the range before the new tree is sequenced, then re-inserted by the InsertAfter code ("code motion" of the local/leaf),

Hm. That code motion is not necessarily safe, and needs to be checked if it's going to be performed. In any case, the `dataLow` operand is not removed from the tree. I may have added this bug when I was working on decomp :(

> In the DecomposeInd case, there is also no sharing, as the "addrBaseHigh" creates a new GT_LCL_VAR reference.

Yup, I missed that. I agree that the `DecomposeInd` case is safe.

pgavlin replied:
> It's more lines of code and less easy to read/reason about, IMO.

I think it's different to reason about, but I personally find it easier: we're in linear IR, so we can insert nodes in any order we like so long as it meets the requirements of data flow & side-effect correctness. Explicit, manual insertion of nodes eliminates any worries that a helper function may be performing erroneous code motion (which is particularly likely in RyuJIT given that many helpers expect to work on tree-ordered HIR on which code motion is easier to perform).

BruceForstall committed:
Respond to CR feedback.

1. Do cheaper circularity and gtPrev checking.
2. Check LIR a bit more frequently.
3. Use gtPrepareCost instead of SeqTree to avoid illegal/dangerous LIR changes.

BruceForstall commented:
@pgavlin I pushed some changes: CR feedback and updates to gtNext/gtPrev checking.

pgavlin commented on src/jit/lir.cpp:
```diff
@@ -1103,12 +1106,51 @@ bool LIR::Range::ContainsNode(GenTree* node) const
+    // Check the gtNext/gtPrev links: (1) ensure there are no circularities, (2) ensure the gtPrev list is
+    // precisely the inverse of the gtNext list.
+    //
+    // To detect circularity, use the "tortoise and hare" 2-pointer algorithm.
```
Neat! I veeeery vaguely remember this algorithm from university.

pgavlin commented:
LGTM
